src/template.rs

```diff
@@ -245,160 +379,334 @@ impl<B: Bundle> IntoIterator for Fragment<B> {
+/// <children> = "=> [" <template> "]"           
+///   <$ident> = an opaque rust identifier
+///    <$expr> = a rust expression of a given type
+///   <$block> = a rust codeblock of a given type
 /// ```
 ///
 #[macro_export]
 macro_rules! template {
-    ($($tail:tt)*) => {{
+    ($($body:tt)*) => {{
         #[allow(unused_mut)]
         let mut fragments = Vec::new();
-        push_item!(fragments; $($tail)*);
+        push_item!(fragments; $($body)*);
         fragments
     }};
 }
 
-/// Used internally. See `template!()`.
+// This pushes the next fragment in the template into the provided vector.
+#[doc(hidden)]
 #[macro_export]
 macro_rules! push_item {
-    // Handle the empty cases.
+    // Empty cases.
+
     () => {};
     ($fragments:ident;) => {};
-    // Handle the case when no name is specified.
-    ($fragments:ident; $block:block $( [ $( $children:tt )+ ] )? ; $( $($sib:tt)+ )?) => {
-        push_fragment!($fragments; { None } $block $( [ $( $children )* ] )* ; $( $( $sib )* )* )
+
+    // Fragments
+
+    // Anonymous bundle expression.
+    ($fragments:ident; $bundle:expr $( => [ $( $children:tt )+ ] )? ; $( $($sib:tt)+ )?) => {
+        push_fragment!($fragments; { None } $bundle $( => [ $( $children )* ] )* ; $( $( $sib )* )* )
     };
-    // Handle the fully specified case, when the name is a stiatic identifier.
-    ($fragments:ident; $name:ident: $block:block $( [ $( $children:tt )+ ] )? ; $( $($sib:tt)+ )?) => {
+    // Bundle expression with fixed name.
+    ($fragments:ident; $name:ident: $bundle:expr $( => [ $( $children:tt )+ ] )? ; $( $($sib:tt)+ )?) => {
         // Stringify the name and throw it in a code-block.
-        push_fragment!($fragments; { Some(stringify!($name).to_string()) } $block $( [ $( $children )* ] )* ; $( $( $sib )* )* )
+        push_fragment!($fragments; { Some(stringify!($name).to_string()) } $bundle $( => [ $( $children )* ] )* ; $( $( $sib )* )* )
     };
-    // Handle the fully specified case, when the name is also a code-block.
-    ($fragments:ident; $name:block: $block:block $( [ $( $children:tt )+ ] )? ; $( $($sib:tt)+ )?) => {
-        push_fragment!($fragments; { Some($name.to_string()) } $block $( [ $( $children )* ] )* ; $( $( $sib )* )* )
+    // Bundle expression with dynamic name.
+    ($fragments:ident; $name:block: $bundle:expr $( => [ $( $children:tt )+ ] )? ; $( $($sib:tt)+ )?) => {
+        push_fragment!($fragments; { Some($name.to_string()) } $bundle $( => [ $( $children )* ] )* ; $( $( $sib )* )* )
     };
-    // Handle the case where it's just a codeblock, returning an iterator of prototypes.
+
+    // Splices
+
+    // A code-block returning an iterator of fragments.
     ($fragments:ident; @ $block:block ; $( $($sib:tt)+ )? ) => {
         $fragments.extend({ $block }); // Extend the fragments with the value of the block.
         $(push_item!($fragments; $($sib)*))* // Continue pushing siblings onto the current list.
     };
```

Comment:
```suggestion
    ($fragments:ident; @ $splice:expr ; $( $($sib:tt)+ )? ) => {
        $fragments.extend({ $splice }); // Extend the fragments with the result of the expression.
        $(push_item!($fragments; $($sib)*))* // Continue pushing siblings onto the current list.
    };
```

Maybe the brace-requirement could be removed for splices as well for consistency:
```rust
@counter(num_sheep, "sheep", Button::Increment, Button::Decrement);
```

Reply:
Forgot to switch from the work account. Again...

Reply:
I considered that, I was still on the fence.